src/utilities/include/par.h

```diff
@@ -49,47 +43,430 @@ inline constexpr ExecutionPolicy autoPolicy(size_t size) {
-      case ExecutionPolicy::Par:                              \
-        return std::NAME(std::execution::par_unseq, args...); \
-      case ExecutionPolicy::Seq:                              \
-        break;                                                \
-    }                                                         \
-    return std::NAME(args...);                                \
-  }
-#define STL_DYNAMIC_BACKEND_VOID(NAME)                 \
-  template <typename... Args>                          \
-  void NAME(ExecutionPolicy policy, Args... args) {    \
-    switch (policy) {                                  \
-      case ExecutionPolicy::Par:                       \
-        std::NAME(std::execution::par_unseq, args...); \
-        break;                                         \
-      case ExecutionPolicy::Seq:                       \
-        std::NAME(args...);                            \
-        break;                                         \
-    }                                                  \
-  }
-#else
-#define STL_DYNAMIC_BACKEND(NAME, RET)             \
-  template <typename Ret = RET, typename... Args>  \
-  Ret NAME(ExecutionPolicy policy, Args... args) { \
-    return std::NAME(args...);                     \
-  }
-#define STL_DYNAMIC_BACKEND_VOID(NAME)              \
-  template <typename... Args>                       \
-  void NAME(ExecutionPolicy policy, Args... args) { \
-    std::NAME(args...);                             \
-  }
-#endif
-
-template <typename... Args>
-void exclusive_scan(ExecutionPolicy policy, Args... args) {
-  // https://github.com/llvm/llvm-project/issues/59810
-  std::exclusive_scan(args...);
+constexpr size_t SEQUENTIAL_LIMIT = 1 << 14;
+
+#if MANIFOLD_PAR == 'T'
+namespace details {
+// implementation from
+// https://duvanenko.tech.blog/2018/01/14/parallel-merge/
+// https://github.com/DragonSpit/ParallelAlgorithms
+// note that the ranges are now [p, r) to fit our convention.
+template <typename T, typename Comp>
+void mergeRec(T *src, T *dest, size_t p1, size_t r1, size_t p2, size_t r2,
+              size_t p3, Comp comp) {
+  size_t length1 = r1 - p1;
+  size_t length2 = r2 - p2;
+  if (length1 < length2) {
+    std::swap(p1, p2);
+    std::swap(r1, r2);
+    std::swap(length1, length2);
+  }
+  if (length1 == 0) return;
+  if (length1 + length2 <= SEQUENTIAL_LIMIT) {
+    std::merge(src + p1, src + r1, src + p2, src + r2, dest + p3, comp);
+  } else {
+    size_t q1 = p1 + length1 / 2;
+    size_t q2 =
+        std::distance(src, std::lower_bound(src + p2, src + r2, src[q1], comp));
+    size_t q3 = p3 + (q1 - p1) + (q2 - p2);
+    dest[q3] = src[q1];
+    tbb::parallel_invoke(
+        [=] { mergeRec(src, dest, p1, q1, p2, q2, p3, comp); },
+        [=] { mergeRec(src, dest, q1 + 1, r1, q2, r2, q3 + 1, comp); });
+  }
+}
+
+template <typename T, typename Comp>
+void mergeSortRec(T *src, T *dest, size_t begin, size_t end, Comp comp) {
```

Comment:
This parallel merge sort is really needed to get performance to comparable level. Sequential sort just cause a blowup in running time for `GenericTwinBooleanTest7081` from 7s to ~30s iirc. And tbb parallel sort is not stable, so it cannot be used.

Also, I know that this is far from optimal, we can probably do much better than this because the parallel `stable_sort` from pstl is faster than ours. But that is a problem for the other day.